File: 2021/10/16/reverse_only_letters.py
"""
https://leetcode.com/problems/reverse-only-letters/
Author: Andrew Jarombek
Date: 10/16/2021
"""


def reverse_only_letters(s: str) -> str:
    start = 0
    end = len(s) - 1

    start_s = ""
    end_s = ""

    while start <= end:
        c = ord(s[start])

        if (c >= 65 and c <= 90) or (c >= 97 and c <= 122):
            swapped = False

            while end != start:
                c_end = ord(s[end])
                end -= 1

                if (c_end >= 65 and c_end <= 90) or (c_end >= 97 and c_end <= 122):
                    start_s += chr(c_end)
                    end_s = chr(c) + end_s
                    swapped = True
                    break
                else:
                    end_s = chr(c_end) + end_s

            if not swapped:
                start_s += chr(c)
        else:
            start_s += chr(c)

        start += 1

    return start_s + end_s

File: 2023/05/15/reverseOnlyLetters.js
/**
 * https://leetcode.com/problems/reverse-only-letters/
 * @author Andrew Jarombek
 * @since 5/15/2023
 */

const assert = require('assert');

const reverseOnlyLetters = function(s) {
    const chars = s.split('');
    let start = 0;
    let end = s.length - 1;

    while (start < end) {
        const c1 = chars[start];
        if (!((c1 >= 'a' && c1 <= 'z') || (c1 >= 'A' && c1 <= 'Z'))) {
            start++;
            continue;
        }

        const c2 = chars[end];

        if (!((c2 >= 'a' && c2 <= 'z') || (c2 >= 'A' && c2 <= 'Z'))) {
            end--;
            continue;
        }

        chars[start] = c2;
        chars[end] = c1;
        start++;
        end--;
    }

    return chars.join('');
};

// Test case 1: Reversing letters in a string
assert.strictEqual(reverseOnlyLetters('ab-cd'), 'dc-ba');

// Test case 2: Reversing letters in a string with mixed characters
assert.strictEqual(reverseOnlyLetters('a-bC-dEf-ghIj'), 'j-Ih-gfE-dCba');
